src/fd.cpp

```diff
@@ -85,38 +93,62 @@ static int wait_thread(void *arg) {
     } else if (waitFor_ms > LATER_POLL_INTERVAL) {
       waitFor_ms = LATER_POLL_INTERVAL;
     }
-    ready = LATER_POLL_FUNC(pollfds.data(), args->num_fds, static_cast<int>(waitFor_ms));
+    ready = LATER_POLL_FUNC(args->fds->data(), args->num_fds, static_cast<int>(waitFor_ms));
     if (args->flag->load()) return 1;
     if (ready) break;
   }
 
   // store pollfd revents in args->results for use by callback
+
   if (ready > 0) {
-    for (int i = 0; i < args->rfds; i++) {
-      (*args->results)[i] = pollfds[i].revents == 0 ? 0 : pollfds[i].revents & POLLIN ? 1: R_NaInt;
-    }
-    for (int i = args->rfds; i < (args->rfds + args->wfds); i++) {
-      (*args->results)[i] = pollfds[i].revents == 0 ? 0 : pollfds[i].revents & POLLOUT ? 1 : R_NaInt;
-    }
-    for (int i = args->rfds + args->wfds; i < args->num_fds; i++) {
-      (*args->results)[i] = pollfds[i].revents != 0;
-    }
-  } else if (ready == 0) {
     for (int i = 0; i < args->num_fds; i++) {
-      (*args->results)[i] = 0;
-    }
-  } else {
-    for (int i = 0; i < args->num_fds; i++) {
-      (*args->results)[i] = R_NaInt;
+      (*args->results)[i] = (*args->fds)[i].revents == 0 ? 0 : (*args->fds)[i].revents & (POLLIN | POLLOUT) ? 1: NA_INTEGER;
     }
+  } else if (ready < 0) {
+    std::fill(args->results->begin(), args->results->end(), NA_INTEGER);
   }
 
-  execLaterNative2(later_callback, static_cast<void *>(argsptr.release()), 0, args->loop);
+  callbackRegistryTable.scheduleCallback(later_callback, static_cast<void *>(argsptr.release()), 0, args->loop);
 
   return 0;
 
 }
 
+void execLater_launch_thread(std::shared_ptr<ThreadArgs> args) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(new std::shared_ptr<ThreadArgs>(args));
+
+  tct_thrd_t thr;
+  if (tct_thrd_create(&thr, &wait_thread, static_cast<void *>(argsptr.release())) != tct_thrd_success)
+    Rcpp::stop("Thread creation failed");
+  tct_thrd_detach(thr);
```

Comment:
Sorry to return to this but I don't think it makes sense to do this from the calling thread. If the child thread terminates first, won't it leak its resources, exactly what you're trying to prevent by calling detach in the first place? Calling it first thing from the child thread seems like it should work every time, the pthread_detach man page shows a thread calling detach on itself. You said you had concerns about the main thread exiting first, I'm not sure why that would matter?